lib/sync/sync-processor.js

```diff
@@ -46,19 +25,86 @@ function syncWithBackend(payload, callback) {
-      syncInfo.syncLoopEnd = new Date().getTime();
-      syncStorage.saveLastSyncDataset(datasetId, globalHash, recordsWithHash, syncInfo, function(err){
-        if (err) {
-          syncUtil.doLog(datasetId, "error", "error when save data to cache. error = " + util.inspect(err));
-        }
-        //we don't need to return the error here as there is no need to retry this request
-        return callback();
+      return cb(null, recordsWithHash);
+    }
+  });
+}
+
+function listDataWithTimeout(params, timeout, cb) {
+  async.race([
+    function(timeoutCallback) {
+      setTimeout(function(){
+        timeoutCallback(new Error('Timeout error occured when try to list data from backend for datasetClient ' + params.datasetClientId));
+      }, timeout);
+    },
+    function(listCallback) {
+      listDatafromBackend(params.datasetId, params.queryParams, params.metaData, listCallback);
+    }
+  ], cb);
+}
+
+function recordProcessTime(startTime, success) {
+  var endTime = new Date().getTime();
+  var totalTime = endTime - startTime;
+  metricsClient.gauge(metrics.KEYS.SYNC_REQUEST_TOTAL_PROCESS_TIME, {success: success}, totalTime);
+}
+
+/**
+ * Perform the sync operation
+ * @param {Object} payload the payload of the sync request
+ * @param {Object} payload.id the id of the dataset client
+ * @param {String} payload.datasetId the id of the dataset that needs to be synced
+ * @param {Object} payload.queryParams the query params that is associated with the dataset
+ * @param {Object} payload.metaData the meta data that is associated with the dataset
+ * @param {Timestamp} payload.startTime when the sync request is created 
+ * @param {Function} callback
+ */
+function syncWithBackend(payload, callback) {
+  var datasetClientId = payload.id;
+  var datasetId = payload.datasetId;
+  var startTime = payload.startTime;
+
+  if (!datasetClientId || !datasetId) {
+    recordProcessTime(startTime, false);
+    syncUtil.doLog(syncUtil.SYNC_LOGGER, "error", "no datasetId value found in sync request payload" + util.inspect(payload));
+    return callback();
+  }
+
+  var queryParams = payload.queryParams || {};
+  var metaData = payload.metaData || {};
+  //we need to add this so that if this sync processor crashed before reaching the end, the scheduler will still be able to push a sync request
+  var expectedTimeout = datasets.getDatasetConfig(datasetId).backendListTimeout * 1000 || 5*60*1000;
+
+  async.waterfall([
+    function setSyncStart(cb) {
+      var syncLoopStartTime = new Date().getTime();
+      syncStorage.updateDatasetClient(datasetClientId, {syncLoopStart: syncLoopStartTime, syncLoopEnd: syncLoopStartTime + expectedTimeout}, function(err){
+        return cb(err);
       });
+    },
+    function listData(cb) {
+      listDataWithTimeout({datasetClientId: datasetClientId, datasetId: datasetId, queryParams: queryParams, metaData: metaData}, expectedTimeout - 2*1000, cb);
```

Comment:
I have probably missed something obvious, but why are we taking away the 2 seconds here?